source/extensions/http/header_validators/envoy_default/path_normalizer.cc

```diff
@@ -0,0 +1,268 @@
+                UhvResponseCodeDetail::get().InvalidUrl};
+
+      case PercentDecodeResult::Normalized:
+        // Valid encoding but outside the UNRESERVED character set. The encoding was normalized to
+        // UPPERCASE and the octet must not be decoded. Copy the normalized encoding.
+        *write++ = *read++;
+        *write++ = *read++;
+        *write++ = *read++;
+        break;
+
+      case PercentDecodeResult::DecodedRedirect:
+        // The encoding was properly decoded but, based on the config, the request should be
+        // redirected to the normalized path.
+        redirect = true;
+        ABSL_FALLTHROUGH_INTENDED;
+      case PercentDecodeResult::Decoded:
+        // The encoding was decoded. Store the decoded octet in the last character of the percent
+        // encoding (read[2]) so it will be processed in the next iteration.
+        read += 2;
+        *read = decode_result.octet();
+      }
+      break;
+    }
+
+    case '.': {
+      // Potential "/./" or "/../" sequence
+      if (*(read + 1) == '/' || (read + 1) == end) {
+        // this is a "./" token.
+        if (prev == '/') {
+          // ignore "/./", jump to next segment (read + 2)
+          read += 2;
+        } else if (prev == '.' && *(write - 2) == '/') {
+          // This is a "../" segment, remove the previous segment by back write up to the previous
+          // slash (write - 2).
+          write -= 2;
+          if (write <= path) {
+            // the full input is: "/.."", this is invalid
+            return {HeaderValidator::RejectOrRedirectAction::Reject,
+                    UhvResponseCodeDetail::get().InvalidUrl};
+          }
+
+          // reset write to overwrite the previous segment
+          while (write > path && *(write - 1) != '/') {
+            --write;
+          }
+
+          // skip the "../" token since it's been handled
+          read += 2;
+        } else {
+          // just a dot within a normal path segment, copy it
+          *write++ = *read++;
+        }
+      } else {
+        // just a dot within a normal path segment, copy it
+        *write++ = *read++;
+      }
+      break;
+    }
+
+    case '/': {
+      if (prev == '/' && !config_.uri_path_normalization_options().skip_merging_slashes()) {
+        // Duplicate slash, merge it
+        ++read;
+      } else {
+        *write++ = *read++;
+      }
+      break;
+    }
+
+    default: {
+      if (testChar(kPathHeaderCharTable, ch)) {
+        // valid path character, copy it
+        *write++ = *read++;
+      } else {
+        // invalid path character
+        return {HeaderValidator::RejectOrRedirectAction::Reject,
+                UhvResponseCodeDetail::get().InvalidUrl};
+      }
+    }
+    }
+  }
+
+  *write = '\0';
+
+  auto normalized_length = static_cast<size_t>(write - path);
+  absl::string_view normalized_path{path, normalized_length};
+  header_map.setPath(normalized_path);
+
+  if (redirect) {
```

Comment:
The logic for redirect is slightly different. Redirect occurs if normalized path is different from the original path. I think here we skip the normalization changes due to slash or dot segment collapsing.

I think we can just check the length of the original path and normalized path and redirect if it is different.

Reply:
I'm slightly confused, can you explain this?

- When should redirect occur?
- How does the [`UNESCAPE_AND_REDIRECT `](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/http/header_validators/envoy_default/v3/header_validator.proto#L52) config option change this behavior?
  - If we only go based comparing the length/value of `path` before and after normalization, then I don't think `UNESCAPE_AND_REDIRECT` is needed.